### PR TITLE
James/malfunction detection functions

### LINKED ISSUE #3090: Malfunction Robot Detection
### Description of the task
We would like the robots to be substituted when the robot status messages stated that the robot is malfunctioning. This functions aims to find the malfunctioning robots and add them to the injured robots list so it can be substituted. Here are the 3 scenarios to be implemented:
1. Break Beam Malfunction
2. Low Battery
3. Overheat, motor faults, and other generic errors

### Acceptance criteria

- [ ] a function that detects errors
- [ ] a set of tests that validates each detection

### Blocked By
#3084

### PR DESCRIPTION
**Pull request by jameschuang2002**
### Description

This PR includes a malfunction detection function that can detect injured robots with error codes, motor over temperature warnings, and break beam malfunction through comparison status messages with the ball.

### Testing Done

Basic unit tests were written for each form of detection.

### Resolved Issues

resolves #3090 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/software/ai/hl/stp/play/play.cpp [hunk 1]
```diff
@@ -155,6 +159,40 @@ std::unique_ptr<TbotsProto::PrimitiveSet> Play::get(
...
+            robots.erase(std::remove(robots.begin(), robots.end(), robot), robots.end());
+
+            auto motion_constraints =
+                buildMotionConstraintSet(world_ptr->gameState(), *auto_sub_tactic);
+            auto primitives = auto_sub_tactic->get(world_ptr);
+            CHECK(primitives.contains(robot.id()))
+                << "Couldn't find a primitive for robot id " << robot.id();
+            auto primitive_proto = primitives[robot.id()]->generatePrimitiveProtoMessage(
+                world_ptr, motion_constraints, obstacle_factory);
+            primitives_to_run->mutable_robot_primitives()->insert(
```
Comment (sauravbanna) [inline src/software/ai/hl/stp/play/play.cpp:199]: nit: a small comment here explaining what these lines are doing could be helpful

File: src/software/ai/hl/stp/play/play.cpp [hunk 2]
```diff
@@ -245,147 +285,14 @@ void Play::updateTactics(const PlayUpdate &play_update)
...
 std::vector<std::string> Play::getState()
 {
     // by default just return the name of the play
     return {objectTypeName(*this)};
 }
+
+std::vector<Robot> Play::getInjuredRobots(const World &world)
+{
+    std::vector<Robot> injured_robots;
+    return injured_robots;
```
Comment (sauravbanna) [inline src/software/ai/hl/stp/play/play.cpp]: shouldn't we be getting the injured robots from the world object?

File: src/software/ai/hl/stp/play/tactic_assignment.cpp [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+    // The Munkres matrix gets solved such that there will be exactly one 0 in every
+    // row and exactly one 0 in every column. All other values will be -1. The 0's
+    // indicate the "workers" and "jobs" (robots and tactics for us) that are most
+    // optimally paired together
+    //
+    // Example matrices:
+    //        -1, 0,-1,         and            0,-1,
+    //         0,-1,-1,                       -1, 0,
+    //        -1,-1, 0,
```
Comment (sauravbanna) [inline src/software/ai/hl/stp/play/tactic_assignment.cpp:99]: explanations are great, makes it a lot easier to understand!

File: src/software/ai/hl/stp/play/tactic_assignment.h [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+#include "software/ai/motion_constraint/motion_constraint_set_builder.h"
+#include "software/ai/navigator/trajectory/trajectory_planner.h"
+#include "software/world/robot.h"
+#include "software/world/world.h"
+
+using TacticVector = std::vector<std::shared_ptr<Tactic>>;
+
+
+/**
+ * Assigns the given tactics to as many of the given robots
```
Comment (sauravbanna) [inline src/software/ai/hl/stp/play/tactic_assignment.h:12]: nit: i think we can mention the hungarian algorithm / a bit more about how the assignment is done here

File: src/software/ai/hl/stp/play/tactic_assignment_test.cpp [hunk 5]
```diff
@@ -0,0 +1,748 @@
...
+    move_tactic_1->updateControlParams(Point(-1, 0), Angle::zero(), 0,
+                                       TbotsProto::MaxAllowedSpeedMode::PHYSICAL_LIMIT);
+
+    TacticVector tactics = {move_tactic_1, stop_tactic_1};
+
+    // Both robots are now closest to move_tactic_1's destination. We do NOT want
+    // robot_0 to be assigned to move_tactic_1, because then robot_1 has to move all the
+    // way around to move_tactic_2. What we expect is that robot_0 will be assigned to
+    // move_tactic_2 and "slide over" to make room for robot_1
+    auto robot_navigation_obstacle_config = ai_config.robot_navigation_obstacle_config();
```
Comment (sauravbanna) [inline src/software/ai/hl/stp/play/tactic_assignment_test.cpp:208]: nit: maybe i'm misreading this but this test only has robot_0 right? where did robot_1 come from?

edit: comment seems to be from a later test

File: src/software/ai/hl/stp/play/tactic_assignment_test.cpp [hunk 6]
```diff
@@ -0,0 +1,748 @@
...
+    auto tup = assignTactics(world, tactics, friendly_team.getAllRobots(),
+                             robot_navigation_obstacle_factory, obstacle_list,
+                             path_visualization);
+
+    auto asst = std::get<2>(tup);
+
+    // move_tactic_1 should be the only Tactic assigned a robot, since stop_tactic_1 is a
+    // lower priority than move_tactic_1 so it should be dropped since there's only 1
+    // robot
+    EXPECT_FALSE(asst.find(move_tactic_1) != asst.end());
```
Comment (sauravbanna) [inline src/software/ai/hl/stp/play/tactic_assignment_test.cpp:223]: the comment mentions move_tactic_1 should be assigned, but it seems like this assertion checks the opposite?

also nit: do we wanna check for ID as well in these assertions?

File: src/software/ai/hl/stp/play/tactic_assignment_test.cpp [hunk 7]
```diff
@@ -0,0 +1,748 @@
...
+                             path_visualization);
+
+    auto asst = std::get<2>(tup);
+
+
+    ASSERT_TRUE(asst.find(move_tactic_1) != asst.end());
+    EXPECT_EQ(asst.find(move_tactic_1)->second, robot_1.id());
+}
+
+// TEST_F(TacticAssignmentTest,
```
Comment (sauravbanna) [inline src/software/ai/hl/stp/play/tactic_assignment_test.cpp:519]: do we wanna keep the below test?

File: src/software/ai/hl/stp/play/tactic_assignment_test.cpp [hunk 8]
```diff
@@ -0,0 +1,748 @@
...
+                             path_visualization);
+
+    auto asst = std::get<2>(tup);
+
+    // Check each tactic is assigned to the intended robot
+    for (unsigned int i = 0; i < tactics.size(); i++)
+    {
+        ASSERT_TRUE(asst.find(tactics[i]) != asst.end());
+        EXPECT_EQ(asst.find(tactics[i])->second, expected_robots_assigned[i].id());
+    }
```
Comment (sauravbanna) [inline src/software/ai/hl/stp/play/tactic_assignment_test.cpp:610]: do we want to add an assertion as well that the other robots have StopTactic assigned to them? looks like we're only checking that the intended robot has MoveTactic assigned

File: src/software/ai/hl/stp/play/tactic_assignment_test.cpp [hunk 9]
```diff
@@ -0,0 +1,748 @@
...
+}
+
+// TEST_F(TacticAssignmentTest, test_assignment_with_tiered_assignment)
```
Comment (sauravbanna) [inline src/software/ai/hl/stp/play/tactic_assignment_test.cpp:613]: do we want to keep this test?

File: src/software/ai/hl/stp/play/tactic_assignment_test.cpp [hunk 10]
```diff
@@ -0,0 +1,748 @@
...
+//                 has_move_1 = true;
+//                 EXPECT_EQ(robot_0.id(), robot);
+//             }
+//         }
+//         EXPECT_TRUE(has_goalie);
+//         EXPECT_TRUE(has_move_1);
+//     }
+// }
+
+TEST_F(TacticAssignmentTest, test_offense_play_with_substitution)
```
Comment (sauravbanna) [inline src/software/ai/hl/stp/play/tactic_assignment_test.cpp:720]: nit: would this be better suited in offense_play_test.py or some other offense play specific test file?

**Review by sauravbanna — CHANGES_REQUESTED** [review]
looks great! tests are thorough, just some minor doc issues

File: src/software/ai/hl/stp/play/tactic_assignment.h [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * @param path_planner_factory The path planner factory
+ * @param world The world
+ * @param tactic_vector The tactic vector
+ * @param robots_to_assign The robots to assign to
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/tactic_assignment.h:17]: missing doc params `obstacle_factory`, `obstacle_list`, `path_visualization`

File: src/software/ai/hl/stp/play/play.h [hunk 12]
```diff
@@ -167,6 +151,8 @@ class Play
     virtual void getNextTactics(TacticCoroutine::push_type& yield,
                                 const WorldPtr& world_ptr) = 0;
 
+    virtual std::vector<Robot> getInjuredRobots(const World& world);
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/play.h]: missing docs

File: src/software/sensor_fusion/sensor_fusion.cpp [hunk 13]
```diff
@@ -401,3 +401,92 @@ void SensorFusion::resetWorldComponents()
...
+            auto power_status_msg = robot_status_msg.power_status();
+            if (power_status_msg.breakbeam_tripped())
+            {
+                auto robot = friendly_team.getRobotById(robot_id);
+                auto dist_vector =
+                    new Vector(ball.value().position().x() - robot->position().x(),
+                               ball.value().position().y() - robot->position().y());
+
+                /* check if the robot has the ball, reference distance 2m */
+                if (dist_vector->length() > 2)
```
Comment (itsarune) [inline src/software/sensor_fusion/sensor_fusion.cpp]: make `2` a static const

File: src/software/ai/hl/stp/play/tactic_assignment_test.cpp [hunk 14]
```diff
@@ -0,0 +1,748 @@
...
+
+class TacticAssignmentTest : public ::testing::Test
+{
+   public:
+    TacticAssignmentTest() : ai_config(TbotsProto::AiConfig()) {}
+
+   protected:
+    void SetUp() override
+    {
+        // Give an explicit seed to STP so that our tests are deterministic
```
Comment (itsarune) [inline src/software/ai/hl/stp/play/tactic_assignment_test.cpp:29]: you can delete this line (line 29)

File: src/software/ai/hl/stp/play/tactic_assignment_test.cpp (hunk 10 above)
Comment (itsarune) [reply src/software/ai/hl/stp/play/tactic_assignment_test.cpp:720]: good point

**Review by itsarune — COMMENTED** [review]
looking good... @sauravbanna had some good questions that you should address (in this PR or the other ones)

File: src/software/ai/hl/stp/play/tactic_assignment_test.cpp (hunk 5 above)
Comment (jameschuang2002) [reply src/software/ai/hl/stp/play/tactic_assignment_test.cpp:208]: this one was addressed in #3031

File: src/software/ai/hl/stp/play/tactic_assignment_test.cpp (hunk 9 above)
Comment (jameschuang2002) [reply src/software/ai/hl/stp/play/tactic_assignment_test.cpp:613]: This was addressed in #3031. I added a new issue and tagged it, it's not really the focus on either PRs so would rather not to deal with it now

File: src/software/ai/hl/stp/play/tactic_assignment_test.cpp (hunk 10 above)
Comment (jameschuang2002) [reply src/software/ai/hl/stp/play/tactic_assignment_test.cpp:720]: We have discussed this possibility while implementing; however, offense_play_test does not offer the flexibility to do tactic assignment manually and do the necessary lower-level validations. I would be happy to change it if there's a way to do it, my initial thought is also to put it in offense_play_test

File: src/software/ai/hl/stp/play/tactic_assignment_test.cpp (hunk 7 above)
Comment (jameschuang2002) [reply src/software/ai/hl/stp/play/tactic_assignment_test.cpp:519]: This was addressed in https://github.com/UBC-Thunderbots/Software/issues/3031. I added a new issue and tagged it, it's not really the focus on either PRs so would rather not to deal with it now

File: src/software/ai/hl/stp/play/play.cpp (hunk 2 above)
Comment (jameschuang2002) [reply src/software/ai/hl/stp/play/play.cpp]: thanks for catching this, I don't think I ever use this one, I deleted it. The major function I was using is in team.cpp.

**Comment by jameschuang2002** [comment]
I think a lot of the comments on tactic_assignment_test.cpp is good; however, since I only ported those from the old stp_test.cpp, I am unsure about what the majority of those are doing. I would be happy to address those but is better to tag it as an issue because the malfunction detection should be the major feature?

File: src/software/sensor_fusion/sensor_fusion.h [hunk 15]
```diff
@@ -48,6 +48,16 @@ class SensorFusion
      */
     std::optional<World> getWorld() const;
 
+    /**
+     * detect injured robots in the team that needs to be substituted and add it to
+     * injuredRobots
+     *
+     * @param robot_status_msgs The robot statuses from the sensor fusion packet
+     * @return the list of robotids of the injured robots
```
Comment (itsarune) [inline src/software/sensor_fusion/sensor_fusion.h:56]: this @return does not match the function definition

File: src/software/sensor_fusion/sensor_fusion.h [hunk 16]
```diff
@@ -167,4 +177,7 @@ class SensorFusion
 
     // The timestamp, in seconds, of the most recently received vision packet
     double last_t_capture;
+
+    // tolerance for break beam tolerance
+    const static int break_beam_tolerance = 2;
```
Comment (itsarune) [inline src/software/sensor_fusion/sensor_fusion.h:182]: constants should be CAPITAL_CASE

**Review by itsarune — COMMENTED** [review]
left a few comments

**Pull request closed without merging**